File: C++/DFS/SurroundedRegions.cpp
// https://leetcode.com/problems/surrounded-regions/

class Solution {
public:
    int m, n;
    int dir[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    
    void dfs(vector<vector<char>>& board, vector<vector<bool>>& visit, int i, int j) {
        if (i <  0 || i >= m || j < 0 || j >= n || board[i][j] != 'O' || visit[i][j]) {
            return;
        }
        board[i][j] = '#';
        visit[i][j] = true;
        
        dfs(board, visit, i, j + 1);
        dfs(board, visit, i, j - 1);
        dfs(board, visit, i + 1, j);
        dfs(board, visit, i - 1, j);
        
    }
    
    void solve(vector<vector<char>>& board) {
        m = board.size();
        n = board[0].size();
        vector<vector<bool>> visit(m, vector<bool> (n, false));
        
        for (int j = 0; j < n; j++) {
            dfs(board, visit, 0, j);
            dfs(board, visit, m - 1, j);
        }
        
        for (int i = 1; i < m - 1; i++) {
            dfs(board, visit, i, 0);
            dfs(board, visit, i, n - 1);
        }
        
        
        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == '#') {
                    board[i][j] = 'O';
                }
                else if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }
        
    }
};

File: Python/BFS/SurroundedRegions.py
# https://leetcode.com/problems/surrounded-regions/

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board:
            return None
        rows = len(board)
        cols = len(board[0])
        visited = set()
        def bfs(i, j):
            process = deque()
            process.append((i, j))
            cur_visited = set()
            cur_visited.add((i, j))
            visited.add((i, j))
            flag = 0
            while process:
                x, y = process.popleft()
                directions = [[x - 1, y], [x + 1, y], [x, y - 1], [x, y + 1]]
                for xx, yy in directions:
                    if xx >= rows or xx < 0 or yy < 0 or yy >= cols:
                        flag = 1
                    else:
                        if (xx, yy) not in visited and board[xx][yy] == 'O':
                            visited.add((xx, yy))
                            process.append((xx, yy))
                            cur_visited.add((xx, yy))
            if flag == 1:
                return
            for x, y in cur_visited:
                board[x][y] = 'X'
            return

        for i in range(rows):
            for j in range(cols):
                if board[i][j] == 'O' and (i, j) not in visited:
                    bfs(i, j)
        return

        """
        Do not return anything, modify board in-place instead.
        """
